File: apex-brain/tools/scrape_trending.py
#!/usr/bin/env python3
"""Scrape GitHub Trending and Trendshift for trending repos."""

import argparse
import json
import re
import sys
import urllib.request
import urllib.error


def scrape_github_trending(language: str = "", since: str = "daily") -> list:
    """Parse GitHub Trending page (no API needed)."""
    url = f"https://github.com/trending/{language}?since={since}"
    req = urllib.request.Request(url, headers={"User-Agent": "ApexBrain/1.0"})
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            html = resp.read().decode()
    except urllib.error.URLError as e:
        return [{"status": "error", "message": str(e)}]

    repos = []
    # Match repo links: /owner/repo pattern inside article tags
    for match in re.finditer(r'<h2 class="h3[^"]*">\s*<a href="(/[^"]+)"', html):
        path = match.group(1).strip()
        parts = path.strip("/").split("/")
        if len(parts) == 2:
            repos.append({"owner": parts[0], "name": parts[1], "url": f"https://github.com{path}"})

    # Extract stars today
    star_matches = re.findall(r'(\d[\d,]*)\s+stars\s+today', html)
    for i, stars in enumerate(star_matches):
        if i < len(repos):
            repos[i]["stars_today"] = int(stars.replace(",", ""))

    return repos[:25]


def scrape_trendshift() -> list:
    """Fetch from Trendshift API."""
    url = "https://trendshift.io/api/repositories?language=&since=daily"
    req = urllib.request.Request(url, headers={"User-Agent": "ApexBrain/1.0"})
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            data = json.loads(resp.read())
        return [{"name": r.get("name"), "owner": r.get("owner"), "stars": r.get("stars"), "description": r.get("description", "")} for r in data[:25]]
    except (urllib.error.URLError, json.JSONDecodeError) as e:
        return [{"status": "error", "source": "trendshift", "message": str(e)}]


def main():
    parser = argparse.ArgumentParser(description="Scrape trending repos")
    parser.add_argument("--source", choices=["github", "trendshift", "all"], default="all")
    parser.add_argument("--language", default="", help="Filter by language (GitHub only)")
    parser.add_argument("--since", default="daily", choices=["daily", "weekly", "monthly"])
    parser.add_argument("--dry-run", action="store_true")
    args = parser.parse_args()

    if args.dry_run:
        print(json.dumps({"status": "dry_run", "source": args.source, "language": args.language}))
        return

    results = {}
    if args.source in ("github", "all"):
        results["github_trending"] = scrape_github_trending(args.language, args.since)
    if args.source in ("trendshift", "all"):
        results["trendshift"] = scrape_trendshift()

    print(json.dumps(results, indent=2))


if __name__ == "__main__":
    main()

File: automations/src/trigger/brain/scrape-trending.ts
import { task } from "@trigger.dev/sdk";

const USER_AGENT = "ApexBrain/1.0";

interface TrendingRepo {
  owner: string;
  name: string;
  url: string;
  stars_today?: number;
  description?: string;
}

async function scrapeGitHubTrending(
  language = "",
  since = "daily"
): Promise<TrendingRepo[]> {
  const url = `https://github.com/trending/${language}?since=${since}`;
  const resp = await fetch(url, { headers: { "User-Agent": USER_AGENT } });
  if (!resp.ok) return [];

  const html = await resp.text();
  const repos: TrendingRepo[] = [];

  // Match repo links inside h2.h3 tags
  const repoPattern = /<h2 class="h3[^"]*">\s*<a href="(\/[^"]+)"/g;
  let match: RegExpExecArray | null;
  while ((match = repoPattern.exec(html)) !== null) {
    const path = match[1].trim();
    const parts = path.replace(/^\//, "").split("/");
    if (parts.length === 2) {
      repos.push({
        owner: parts[0],
        name: parts[1],
        url: `https://github.com${path}`,
      });
    }
  }

  // Extract stars today
  const starPattern = /([\d,]+)\s+stars\s+today/g;
  let starMatch: RegExpExecArray | null;
  let i = 0;
  while ((starMatch = starPattern.exec(html)) !== null && i < repos.length) {
    repos[i].stars_today = parseInt(starMatch[1].replace(/,/g, ""), 10);
    i++;
  }

  return repos.slice(0, 25);
}

async function scrapeTrendshift(): Promise<TrendingRepo[]> {
  const url =
    "https://trendshift.io/api/repositories?language=&since=daily";
  const resp = await fetch(url, { headers: { "User-Agent": USER_AGENT } });
  if (!resp.ok) return [];

  const data: any[] = await resp.json();
  return data.slice(0, 25).map((r) => ({
    owner: r.owner,
    name: r.name,
    url: `https://github.com/${r.owner}/${r.name}`,
    stars_today: r.stars,
    description: r.description || "",
  }));
}

export const scrapeTrending = task({
  id: "scrape-trending",
  run: async (payload: {
    source?: "github" | "trendshift" | "all";
    language?: string;
    since?: "daily" | "weekly" | "monthly";
  }) => {
    const source = payload.source || "all";
    const results: Record<string, TrendingRepo[]> = {};

    if (source === "github" || source === "all") {
      results.github_trending = await scrapeGitHubTrending(
        payload.language || "",
        payload.since || "daily"
      );
    }
    if (source === "trendshift" || source === "all") {
      results.trendshift = await scrapeTrendshift();
    }

    console.log(
      `Scraped trending: ${Object.entries(results)
        .map(([k, v]) => `${k}=${v.length}`)
        .join(", ")}`
    );
    return results;
  },
});
